zebrad/src/components/mempool.rs

```diff
@@ -365,11 +365,39 @@ impl Service<Request> for Mempool {
             if let Some(TipAction::Grow { block }) = tip_action {
                 tracing::trace!(block_height = ?block.height, "handling blocks added to tip");
 
+                let spent_outpoints = block
+                    .transactions
+                    .iter()
+                    .flat_map(|tx| tx.spent_outpoints())
+                    .collect();
+
+                let sprout_nullifiers = block
+                    .transactions
+                    .iter()
+                    .flat_map(|transaction| transaction.sprout_nullifiers())
+                    .collect();
+                let sapling_nullifiers = block
+                    .transactions
+                    .iter()
+                    .flat_map(|transaction| transaction.sapling_nullifiers())
+                    .collect();
+                let orchard_nullifiers = block
+                    .transactions
+                    .iter()
+                    .flat_map(|transaction| transaction.orchard_nullifiers())
+                    .collect();
```

Comment:
It looks like the methods `remove_same_effects` and `reject_invalidated_transactions` could be merged, and this code could be moved to the unified method.

Reply:
It would be good to move these collect statements into the new method, to keep poll_ready() short and readable.

Reply:
Good points. Prevents repetition in tests too.